src/timeevolution/timeseries.jl

```diff
@@ -0,0 +1,323 @@
+
+
+"""
+    construct(ct, poss, vels [, ω [, dt=0.01]]) -> xt, yt, vxt, vyt, t
+Given the output of [`evolve!`](@ref), construct the
+timeseries of the position and velocity, as well as the time vector.
+In case of not given `ω` (or `ω == 0`), straight construction takes place.
+In case of `ω != 0` or `ω::Vector` magnetic construction takes place.
+The additional optional argument of `dt` (only valid for magnetic construction)
+is the timestep with which the timeseries are constructed.
+Return:
+* x position time-series
+* y position time-series
+* x velocity time-series
+* y velocity time-series
+* time vector
+"""
+function construct(t::Vector{T},
+    poss::Vector{SVector{2,T}}, vels::Vector{SVector{2,T}}) where {T}
+
+    xt = [pos[1] for pos in poss]
+    yt = [pos[2] for pos in poss]
+    vxt = [vel[1] for vel in vels]
+    vyt = [vel[2] for vel in vels]
+    ct = cumsum(t)
+    return xt, yt, vxt, vyt, ct
+end
+
+function construct(t::Vector{T},  poss::Vector{SVector{2,T}},
+vels::Vector{SVector{2,T}}, ω::T, dt=0.01) where {T}
+
+    dt = T(dt)
+    ω == 0 && return construct(t, poss, vels)
+
+    xt = [poss[1][1]]
+    yt = [poss[1][2]]
+    vxt= [vels[1][1]]
+    vyt= [vels[1][2]]
+    ts = [t[1]]
+    ct = cumsum(t)
+
+    for i in 2:length(t)
+        φ0 = atan(vels[i-1][2], vels[i-1][1])
+        x0 = poss[i-1][1]; y0 = poss[i-1][2]
+        colt=t[i]
+
+        t0 = ct[i-1]
+        # Construct proper time-vector
+        if colt >= dt
+            timevec = collect(0:dt:colt)[2:end]
+            timevec[end] == colt || push!(timevec, colt)
+        else
+            timevec = colt
+        end
+
+        for td in timevec
+            s, c = sincos(ω*td + φ0)
+            s0, c0 = sincos(φ0)
+            push!(vxt, c)
+            push!(vyt, s)
+            push!(xt, s/ω + x0 - s0/ω)  #vy0 is sin(φ0)
+            push!(yt, -c/ω + y0 + c0/ω) #vx0 is cos(φ0)
+            push!(ts, t0 + td)
+        end#collision time
+    end#total time
+    return xt, yt, vxt, vyt, ts
+end
+
+
+
+"""
+    extrapolate(p, prevpos, prevvel, ct, dt)
+one day, this will be a beautiful docstring
+"""
+function extrapolate(p::MagneticParticle{T}, prevpos::SV{T}, prevvel::SV{T}, ct::T,
+                     dt::T) where {T <: AbstractFloat}
+    poss = SV{T}[]
+    vels = SV{T}[]
+    
+    φ0 = atan(prevvel[2], prevvel[1])
+    s0, c0 = sincos(φ0)
+    sc0 = SV(s0, -c0)
+
+    tvec = collect(0:dt:ct)
+    
+    for t ∈ tvec
+        s,c = sincos(p.ω*t + φ0)
+
+        push!(poss, SV(s,-c)/p.ω + prevpos - sc0/p.ω)
+        push!(vels, SV(s,c))        
+    end
+
+    # finish with ct
+    if tvec[end] != ct
+        push!(tvec, ct)
+        push!(poss, p.pos + p.current_cell)
+        push!(vels, p.vel)
+    end
+    
+    return poss, vels, tvec
+end
+
+function extrapolate(p::Particle{T}, prevpos::SV{T}, prevvel::SV{T}, ct::T,
+                     dt::T) where {T <: AbstractFloat}
+
+    poss = SV{T}[]
+    vels = SV{T}[]
+
+    tvec = collect(0:dt:ct)
```

Comment:
Its clear from here that the length of `poss` and `vels` etc. is known exactly. It is much better to initialize them with proper length and zeros than to `push!` element by element.

At the end you can push the single final element if need be.